### PR TITLE
Partitioned Epoch Reward Distribution

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

prototype of https://github.com/solana-labs/solana/pull/27455

#### Summary of Changes

Fixes #

### CODE REVIEW

File: ledger/src/blockstore_processor.rs [hunk 1]
```diff
@@ -1319,6 +1319,19 @@ fn process_next_slots(
     leader_schedule_cache: &LeaderScheduleCache,
     pending_slots: &mut Vec<(SlotMeta, Bank, Hash)>,
 ) -> result::Result<(), BlockstoreProcessorError> {
+    if bank.start_epoch_reward_calc() {
```
Comment (HaoranYi) [inline ledger/src/blockstore_processor.rs:1322]: @sakridge

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -3054,54 +3391,337 @@ impl Bank {
...
+        for x in &vote_rewards[begin..end] {
+            if let Some(vote_reward_info) = x.vote_reward_info {
+                rewards.push((x.vote_pubkey, vote_reward_info));
+                num_vote_rewards += 1;
+            }
+        }
+        (num_stake_rewards, num_vote_rewards)
+    }
+
+    fn credit_epoch_rewards_in_partition(&mut self, start_slot: Slot, partition_index: u64) {
```
Comment (HaoranYi) [inline runtime/src/bank.rs:3651]: @jeffwashington This fn stores the accounts in multiple slots.

File: ledger/src/blockstore_processor.rs (hunk 1 above)
Comment (HaoranYi) [reply ledger/src/blockstore_processor.rs:1322]: ah. this is only for ledger replay.

**Comment by t-nelson** [comment]
we're going to need to split this up.  it's much too large to consider as is

File: frozen-abi/src/abi_example.rs [hunk 3]
```diff
@@ -561,3 +561,12 @@ impl<T: AbiExample> AbiExample for once_cell::sync::OnceCell<T> {
         Self::with_value(T::example())
     }
 }
+
+use crossbeam_channel::{unbounded, Sender};
+
+impl<T: AbiExample> AbiExample for Sender<T> {
```
Comment (t-nelson) [inline frozen-abi/src/abi_example.rs:567]: what change made this a requirement?  a channel should definitely _not_ be part of ABI

File: runtime/src/epoch_reward_calc_service.rs [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+
+            calc_metrics.total_time_us = total_time.as_us();
+            calc_metrics.report();
+        }
+    }
+
+    /// Handle incoming calculation requests loop
+    pub fn handle_request(&self) {
+        const MAX_REQUEST_RECV_TIME_MS: u64 = 500;
+        if let Ok(request) = self
```
Comment (t-nelson) [inline runtime/src/epoch_reward_calc_service.rs:289]: this should be handled directly in the loop so we can break on error.  discarding the error like this will very likely cause the app to hang on exit and/or allow the service to silently die in the background.  neither of which are desirable operationally

File: runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs [hunk 5]
```diff
@@ -230,7 +230,10 @@ impl SnapshotStorageRebuilder {
     ) {
         thread_pool.spawn(move || {
             for path in rebuilder.file_receiver.iter() {
-                rebuilder.process_append_vec_file(path).unwrap();
+                let r = rebuilder.process_append_vec_file(path);
+                if let Err(e) = r {
+                    debug!("process append vec slot complete failure: {e}");
+                }
```
Comment (t-nelson) [inline runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:236]: these changes seem unrelated

**Review by t-nelson — COMMENTED** [review]
i haven't looked at much of the tests yet, nor whatever got pushed today.

a few broad comments
* always prefer descriptive names to short ones (exception: one-liner closures)
* especially in runtime, avoid implicit symbolic math operators in favor of the explicit, behavior-declaring methods.
* avoid `unwrap`ing unless the operation is known infallible. add a comment to justify it
* prefer `ToType::{try_,}from(from_var)` over `from_var as ToType` or `from_var.{try_,}into()`. it better self-documents and doesn't have strange side-effects on primitives

  > it _may_ be time for a style guide :sweat_smile: 

as for places to break up this PR, i've named a few in the comments, but will reiterate here
* the rpc service changes can go in separately
* the changes to the stake program can go in separately
* i believe the helpers added to `VoteState` and the logic using them can go in separately (behavior should be the same today as with the partitioned rewards)
* if the new `Bank` fields end up needing to go into snapshots, we should break their addition out and add them inertly

more opportunities to break this monster up should show themselves as we get through a few more rounds of review
